### PR TITLE
Contact Info & Map widget: Add legacy widget → block transform

### PR DESCRIPTION
**Pull request by ivan-ottinger**
#### Changes proposed in this Pull Request:
This PR adds transform for the legacy / classic "Contact Info & Map" widget.

Markup on 2021-09-29 at 14:34:20

The proposed change can help users move the content of their existing widget into the "Contact Info" block, plus several additional blocks:

- Heading (H2) block storing the `Title` of the widget
- Paragraph block storing the `Hours` field of the widget

At the moment, the `Address` field of the widget is transformed into the `Street Address` field of the block, leaving other block fields (e.g. Country, Postal/Zip Code, State) empty. We may consider parsing the `Address` field of the widget into separate block fields in the future iteration.

#### Jetpack product discussion
https://github.com/Automattic/jetpack/issues/21069

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Add the "Contact Info & Map" widget (using the Legacy Block) to your testing site and save it.
2. Select the transform option and choose "Contact Info" block (as can be seen in the screenshot above).
3. The widget information should be correctly transformed.

### CODE REVIEW

**Comment by ivan-ottinger** [comment]
There's an issue which is blocking the build process of the WPCOM patch: p1632857344442400-slack-CBG1CP4EN - so we cannot test the patch on WPCOM yet.

File: projects/plugins/jetpack/modules/widgets/contact-info.php [hunk 1]
```diff
@@ -31,6 +31,7 @@ public function __construct() {
 				'classname'                   => 'widget_contact_info',
 				'description'                 => __( 'Display a map with your location, hours, and contact information.', 'jetpack' ),
 				'customize_selective_refresh' => true,
+				'show_instance_in_rest'       => true,
 			);
```
Comment (arcangelini) [inline projects/plugins/jetpack/modules/widgets/contact-info.php:35]: Not sure if you left it out on purpose, but the hook to remove it from the Legacy block needs to be added as well. Made testing easier though : )

File: projects/plugins/jetpack/extensions/blocks/contact-info/index.js [hunk 2]
```diff
@@ -51,6 +52,50 @@ export const settings = {
...
+						} ),
+						createBlock( 'jetpack/phone', {
+							phone: instance.raw.phone,
+						} ),
+						createBlock( 'jetpack/address', {
+							address: instance.raw.address,
+						} ),
+					];
+
+					if ( instance.raw.hours ) {
```
Comment (arcangelini) [inline projects/plugins/jetpack/extensions/blocks/contact-info/index.js:83]: You might need to make all of the innerBlocks conditional. Otherwise, you end up with blanks blocks if it is empty. This is what it transformed to for me with certain fields blank - https://d.pr/i/8UNAOy

File: projects/plugins/jetpack/extensions/blocks/contact-info/index.js [hunk 3]
```diff
@@ -51,6 +52,50 @@ export const settings = {
...
+						innerBlocks = [
+							...innerBlocks,
+							createBlock( 'core/paragraph', { content: instance.raw.hours } ),
+						];
+					}
+
+					if ( instance.raw.showmap ) {
```
Comment (arcangelini) [inline projects/plugins/jetpack/extensions/blocks/contact-info/index.js:90]: You could also refractor this to all be in the array? Something like:

```
...( instance.raw.phone ? [ createBlock( 'jetpack/phone', { phone: instance.raw.hours } ) ] : [] ),
...( instance.raw.hours ? [ createBlock( 'core/paragraph', { content: instance.raw.hours } ) ] : [] ),
...( instance.raw.showmap ? [ createBlock( 'jetpack/map' ) ] : [] ),

```

**Review by arcangelini — CHANGES_REQUESTED** [review]
Everything worked for me, just need to add the hook to hide the widget and possibly the conditionals for the other fields?

File: projects/plugins/jetpack/modules/widgets/contact-info.php (hunk 1 above)
Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: Thanks! :) 

Yes, I left in out on purpose - as I wanted to hear your opinions on also leaving the Contact Info & Map widget there for the users to choose (in case they prefer it over Contact Info block).

File: projects/plugins/jetpack/extensions/blocks/contact-info/index.js (hunk 2 above)
Comment (ivan-ottinger) [reply projects/plugins/jetpack/extensions/blocks/contact-info/index.js:83]: I was initially thinking about making all fields conditional, but then decided to leave some of them unconditional. The reasoning here is that if the user is transforming to the Contact Info block, they will see what fields they can fill in (even if they didn't have them filled in the widget previously).

Also, for example, if the widget title is empty, they will see a nice empty header they can fill in. At least this is how I would like it to be (if I were a user). But I get your point!

Comment (arcangelini) [reply projects/plugins/jetpack/extensions/blocks/contact-info/index.js:83]: I totally see your point. Generally speaking, I imagine the transform being used for someone who already has a fully built widget and we are just preserving/transferring the settings to the new format. In that case I imagine it could be unnecessary.

It could go either way I guess haha.

File: projects/plugins/jetpack/modules/widgets/contact-info.php (hunk 1 above)
Comment (arcangelini) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: My understanding was that we were getting rid of the redundancies in the widget area. I think it is a little confusing to users to have two or three options for the same things. This seemed like a 1:1 transform, does it differ greatly between the widget and block option?

I am not completely sure though, it could be best to get @simison and @yansern opinion.

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: >  I think it is a little confusing to users to have two or three options for the same things.

I see what you mean by looking at this from the perspective of user's confusion.

> This seemed like a 1:1 transform, does it differ greatly between the widget and block option?

The main difference I can see is the map solution being used: Google Maps (Contact Info & Map widget) vs MapBox (when the user transforms the widget into blocks).

I personally prefer MapBox, but I can see some users that would like to use Google Maps instead.

If we want to remove the redundancy, then yes, I think it would be best to remove the legacy widget.

Comment (yansern) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: I see that the only way to use Google Maps is through shortcodes.
`[googlemaps https://maps/url/here]`

1. Are we able to construct a Google Maps embed URL with long & lat param?
2. Is it alright if it's a shortcode block?

Reference: https://wordpress.com/support/google-maps/#embedding-a-google-map

### How about creating a block for this?

Also, another aspect we can consider, is to create a Contact Info & Maps block **as part of our Gutenberg training**. Since the widget code lives in Jetpack, why not create the same block in Jetpack?

Then, we can come back to this transformation exercise and have it do a 1:1 transformation.

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: > Are we able to construct a Google Maps embed URL with long & lat param?

For our use case, I think we could construct the Google Map embed URL using information from the physical address that the user provides, e.g. as we can see in [this example code](https://developers.google.com/maps/documentation/embed/get-started):

`https://www.google.com/maps/embed/v1/place?key=API_KEY&q=Space+Needle,Seattle+WA`

> Is it alright if it's a shortcode block?

I would personally rather moved away from shortcode if possible and voted for a block solution.

We could also try to utilize the existing Map block and use [forward geocoding](https://docs.mapbox.com/api/search/geocoding/#forward-geocoding) to get the necessary coordinates for the map.

Comment (jeherve) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: > The main difference I can see is the map solution being used: Google Maps (Contact Info & Map widget) vs MapBox (when the user transforms the widget into blocks).

This is indeed a big difference. Apart from the look of each service, the requirements, pricing, and API setup are also different. 

Switching from one to the next is not straightforward on self-hosted sites, where folks will need to have an API setup step (that's not necessary on WordPress.com where we handle both services for them; there switching from one service to the next is indeed straightforward).

I wonder if we should take that into account in general; if someone has a Contact Info Widget already set up, with a map, then we know what their requirements are. We could then check if they also have a Mapbox API setup on that site, and offer the transform if we know that's possible for them.

What do you think?

> I would personally rather moved away from shortcode if possible and voted for a block solution.

+1.

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: > This is indeed a big difference. Apart from the look of each service, the requirements, pricing, and API setup are also different.

> Switching from one to the next is not straightforward on self-hosted sites [...]

That's a good point. 

> [...] if someone has a Contact Info Widget already set up, with a map, then we know what their requirements are. We could then check if they also have a Mapbox API setup on that site, and offer the transform if we know that's possible for them.

> What do you think?

I think we could create a new block that would be an equivalent to the existing widget - and rely on Google Maps initially. This would let us move faster with the retirement of legacy widgets.

Then, we could add a dropdown to the new block settings where the user would be able to choose whether they want to use Google Maps or Mapbox.

This would provide flexibility when it comes to block transform. As you suggested, we could then also check whether the user has the Mapbox API set up already and if yes, we could transform the legacy widget to the new block with Mapbox map.

Even if we decide not to implement Mapbox into the new block (and leave Google Maps support there only), I would consider it as a good outcome - as the users would still have the option to use the Mapbox block separately. I think this would also make the block transform more straightforward - as we would be transforming widget into block 1:1 (the same fields would be available in the block as they currently are in the widget).

In other words, as a first iteration we could have a block that has the same functionality as the current legacy widget. Then we could build on top of that and add extra features, e.g. Mapbox support.

How does that sound? 🙂

Comment (yansern) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: > In other words, as a first iteration we could have a block that has the same functionality as the current legacy widget. Then we could build on top of that and add extra features, e.g. Mapbox support.

I agree with starting simple. Showing the same form as what the legacy widget did, and use Google Maps, would be the path to least resistance. I imagine literally extract bits & pieces and put them back using Block's library.

Comment (yansern) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: It seems that there's a Contact Info block on Jetpack. It also creates a Google Maps.
Perhaps the work needed is to modify this block and give user the option to embed the Google Maps under.
https://github.com/Automattic/jetpack/tree/master/projects/plugins/jetpack/extensions/blocks/contact-info

Comment (jeherve) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: > How does that sound? 🙂

Yeah, we could give that a try. I would, however, suggest that you post on a P2 about this as there are some product discussions to be had about this. See p7fD6U-2XB-p2 for some background.

> It also creates a Google Maps

It's just a link though, so nothing super fancy. But I agree we should start there with that block, maybe by leveraging the existing map block as a new possible inner block for that block.

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: I have just publish a P2 post to gather more insights so we can then decide on the best approach moving forward: p7fD6U-4KX-p2.

Comment (jeherve) [reply projects/plugins/jetpack/modules/widgets/contact-info.php:35]: Discussion continues in p7fD6U-4KX-p2
